File: coding-test/javascript/src/kakao-internship-2020/keypad-press.js
// @ts-check

/**
 * @param {number[]} numbers
 * @param {string} hand
 * @returns {string}
 */
function solution(numbers, hand) {
  let answer = ''

  const rows = [0, 3, 3, 3, 2, 2, 2, 1, 1, 1]
  const leftHand = [0, 0]
  const rightHand = [0, 2]

  for (let i = 0; i < numbers.length; i += 1) {
    const number = numbers[i]

    if (number === 1 || number === 4 || number === 7) {
      answer += 'L'
      leftHand[0] = rows[number]
      leftHand[1] = 0
    }

    if (number === 3 || number === 6 || number === 9) {
      answer += 'R'
      rightHand[0] = rows[number]
      rightHand[1] = 2
    }

    if (number === 0 || number === 2 || number === 5 || number === 8) {
      const left =
        Math.abs(leftHand[0] - rows[number]) + Math.abs(leftHand[1] - 1)
      const right =
        Math.abs(rightHand[0] - rows[number]) + Math.abs(rightHand[1] - 1)

      if (left < right || (left === right && hand === 'left')) {
        answer += 'L'
        leftHand[0] = rows[number]
        leftHand[1] = 1
      } else if (left > right || (left === right && hand === 'right')) {
        answer += 'R'
        rightHand[0] = rows[number]
        rightHand[1] = 1
      }
    }
  }

  return answer
}

function diffSolution(numbers, hand) {
  hand = hand[0] === 'r' ? 'R' : 'L'
  const position = [1, 4, 4, 4, 3, 3, 3, 2, 2, 2]
  const h = {
    L: [1, 1],
    R: [1, 1],
  }
  return numbers
    .map((x) => {
      if (/[147]/.test(x)) {
        h.L = [position[x], 1]
        return 'L'
      }

      if (/[369]/.test(x)) {
        h.R = [position[x], 1]
        return 'R'
      }

      const distL = Math.abs(position[x] - h.L[0]) + h.L[1]
      const distR = Math.abs(position[x] - h.R[0]) + h.R[1]
      if (distL === distR) {
        h[hand] = [position[x], 0]
        return hand
      }

      if (distL < distR) {
        h.L = [position[x], 0]
        return 'L'
      }

      h.R = [position[x], 0]
      return 'R'
    })
    .join('')
}

const numbers = [1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5]
const hand = 'right'
const answer = 'LRLLLRLLRRL'

console.log(answer === solution(numbers, hand))
console.log(answer === diffSolution(numbers, hand))

File: practice-coding-test/java-kotlin/src/kr/co/programmers/java/kakao_internship_2020/KeypadPress.java
package kr.co.programmers.java.kakao_internship_2020;

public class KeypadPress {

    public static void main(String[] args) {
        int[] numbers = {1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5};
        String hand = "right";
        System.out.println(new Solution().solution(numbers, hand));
    }

    static class Solution {
        public String solution(int[] numbers, String hand) {

            StringBuilder answer = new StringBuilder();

            String leftLineNumbers = "147";
            String rightLineNumbers = "369";
            String middleLineNumbers = "2580";

            int[] leftHand = {0, 3};
            int[] rightHand = {2, 3};

            for(int i = 0; i < numbers.length; i++){

                String num = String.valueOf(numbers[i]);

                if(leftLineNumbers.contains(num)){
                    leftHand[0] = 0;
                    leftHand[1] = leftLineNumbers.indexOf(num);
                    answer.append("L");
                    continue;
                }

                if(rightLineNumbers.contains(num)){
                    rightHand[0] = 2;
                    rightHand[1] = rightLineNumbers.indexOf(num);
                    answer.append("R");
                    continue;
                }

                int leftHandDistance = Math.abs(leftHand[0] - 1) + Math.abs(leftHand[1] - middleLineNumbers.indexOf(num));
                int rightHandDistance = Math.abs(rightHand[0] - 1) + Math.abs(rightHand[1] - middleLineNumbers.indexOf(num));

                int difference = leftHandDistance - rightHandDistance;

                if(difference > 0 || (difference == 0 && "right".equals(hand))){
                    rightHand[0] = 1;
                    rightHand[1] = middleLineNumbers.indexOf(num);
                    answer.append("R");
                } else if(leftHandDistance - rightHandDistance < 0 || (difference == 0 && "left".equals(hand))){
                    leftHand[0] = 1;
                    leftHand[1] = middleLineNumbers.indexOf(num);
                    answer.append("L");
                }

            }

            return answer.toString();
        }

    }
}
